pipeline_dp/dp_engine.py

```diff
@@ -105,6 +107,76 @@ def _aggregate(self, col, params: AggregateParams,
 
         return col
 
+    def select_private_partitions(self, col,
+                                  params: SelectPrivatePartitionsParams,
+                                  data_extractors: DataExtractors):
+        """Retrieves a list of differentially-private partitions.
+
+        Args:
+          col: collection with elements of the same type.
+          params: parameters, see doc for SelectPrivatePartitionsParams.
+          data_extractors: functions that extract needed pieces of information
+            from elements of 'col'. Only privacy_id_extractor and partition_extractor are required.
+            value_extractor is not required.
+        """
+        self._report_generators.append(ReportGenerator(params))
+        max_partitions_contributed = params.max_partitions_contributed
+
+        # Extract the columns.
+        col = self._ops.map(
+            col, lambda row: (data_extractors.privacy_id_extractor(row),
+                              data_extractors.partition_extractor(row)),
+            "Extract (privacy_id, partition_key))")
+        # col : (privacy_id, partition_key)
+
+        # Apply cross-partition contribution bounding
+        col = self._ops.group_by_key(col)
+
+        # col : (privacy_id, [partition_key])
+
+        # Note: This may not be scalable if a single privacy ID contributes
+        # to _way_ too many partitions.
+        def unique_elements_fn(pid_and_pks):
+            pid, pks = pid_and_pks
+            unique = set()
+            for pk in pks:
+                unique.add(pk)
+
+            return ((pid, pk) for pk in unique)
```

Comment:
Nit: it might be done simpler:
```
pid, pks = pid_and_pks
return ((pid, pk) for pk in set(pks))
```

But even better, we can do sampling in this fn (calling `sample_fixed_per_key` is one more group_by, which is slow). Sampling might be easily done with [np.random.choice](https://numpy.org/doc/stable/reference/random/generated/numpy.random.choice.html)

Reply:
Nice, this allowed to remove a few stages!